File: modulo-diario/api/app/schemas/matter.py
import uuid
from datetime import date, datetime

from pydantic import BaseModel, field_validator

from app.core.html_sanitizer import sanitize_html
from app.models.enums import MatterStatus

PUBLICATION_TYPES = ("normal", "rectification", "republication")


class MatterCreate(BaseModel):
    title: str
    summary: str | None = None
    act_type_id: uuid.UUID
    org_unit_id: uuid.UUID | None = None
    content_html: str
    content_json: dict | None = None
    content_mode: str = "rich_text"
    act_number: str | None = None
    act_year: int | None = None
    act_date: date | None = None
    responsible_name: str | None = None
    responsible_role: str | None = None
    responsible_id: uuid.UUID | None = None
    metadata: dict | None = None
    publication_type: str = "normal"
    references_matter_id: uuid.UUID | None = None

    @field_validator("title")
    @classmethod
    def title_not_empty(cls, v: str) -> str:
        v = v.strip()
        if not v:
            raise ValueError("Title cannot be empty")
        return v

    @field_validator("content_html")
    @classmethod
    def sanitize(cls, v: str) -> str:
        return sanitize_html(v)


class MatterUpdate(BaseModel):
    title: str | None = None
    summary: str | None = None
    act_type_id: uuid.UUID | None = None
    org_unit_id: uuid.UUID | None = None
    content_html: str | None = None
    content_json: dict | None = None
    content_mode: str | None = None
    act_number: str | None = None
    act_year: int | None = None
    act_date: date | None = None
    responsible_name: str | None = None
    responsible_role: str | None = None
    responsible_id: uuid.UUID | None = None
    metadata: dict | None = None
    publication_type: str | None = None
    references_matter_id: uuid.UUID | None = None

    @field_validator("title")
    @classmethod
    def title_not_empty(cls, v: str | None) -> str | None:
        if v is not None:
            v = v.strip()
            if not v:
                raise ValueError("Title cannot be empty")
        return v

    @field_validator("content_html")
    @classmethod
    def sanitize(cls, v: str | None) -> str | None:
        if v is not None:
            return sanitize_html(v)
        return v


class MatterReviewDecision(BaseModel):
    """Payload for reviewer actions that require an explanation (reject)."""
    reason: str | None = None
    action: str | None = None  # informational: "approve" | "reject"


class MatterRectifyRequest(BaseModel):
    """Optional overrides when creating a rectification of a published matter."""

    title: str | None = None
    summary: str | None = None
    content_html: str | None = None
    content_json: dict | None = None
    notes: str | None = None


class AttachmentOut(BaseModel):
    id: uuid.UUID
    file_id: uuid.UUID
    title: str | None
    type: str
    position: int

    model_config = {"from_attributes": True}


class MatterResponse(BaseModel):
    id: uuid.UUID
    title: str
    summary: str | None
    act_type_id: uuid.UUID
    org_unit_id: uuid.UUID | None
    content_html: str
    content_json: dict | None
    content_mode: str
    plain_text: str
    status: MatterStatus
    workflow_status: str | None = None
    version: int
    author_id: uuid.UUID
    reviewed_by: uuid.UUID | None
    published_at: datetime | None
    is_erratum: bool
    act_number: str | None = None
    act_year: int | None = None
    act_date: date | None = None
    responsible_name: str | None = None
    responsible_role: str | None = None
    responsible_id: uuid.UUID | None = None
    metadata: dict | None = None
    review_reason: str | None = None
    publication_type: str = "normal"
    references_matter_id: uuid.UUID | None = None
    slug: str | None = None
    created_at: datetime
    updated_at: datetime
    attachments: list[AttachmentOut] = []

    model_config = {"from_attributes": True}


class MatterListResponse(BaseModel):
    id: uuid.UUID
    title: str
    summary: str | None
    act_type_id: uuid.UUID
    org_unit_id: uuid.UUID | None
    status: MatterStatus
    workflow_status: str | None = None
    version: int
    author_id: uuid.UUID
    reviewed_by: uuid.UUID | None
    act_number: str | None = None
    act_year: int | None = None
    slug: str | None = None
    created_at: datetime
    updated_at: datetime
    attachment_count: int = 0

    model_config = {"from_attributes": True}


class MatterWorkflowUpdate(BaseModel):
    """Avança/retorna o fluxo de criação do documento (não altera o status
    editorial)."""

    status: str
    note: str | None = None


class MatterWorkflowHistoryOut(BaseModel):
    id: uuid.UUID
    action: str
    description: str | None = None
    from_status: str | None = None
    to_status: str | None = None
    user_id: uuid.UUID | None = None
    created_at: datetime


class MatterNextTitleResponse(BaseModel):
    title: str
    next_number: int
    last_number: int
    year: int
    # The number is only a suggestion computed from already-approved/published
    # matters. It is NEVER a reservation — two authors can receive the same
    # suggestion. UI must label it "número sugerido".
    advisory: bool = True
    reserved: bool = False


class MessageResponse(BaseModel):
    message: str

File: modulo-diario/web-admin/src/types/matter.ts
export type MatterStatus =
  | "draft"
  | "review"
  | "approved"
  | "published"
  | "archived"
  | "rejected";

export interface MatterListItem {
  id: string;
  title: string;
  summary: string | null;
  act_type_id: string;
  org_unit_id: string | null;
  status: MatterStatus;
  version: number;
  author_id: string;
  reviewed_by: string | null;
  created_at: string;
  updated_at: string;
  attachment_count: number;
}

export interface Attachment {
  id: string;
  file_id: string;
  title: string | null;
  type: string;
  position: number;
}

export interface Matter {
  id: string;
  title: string;
  summary: string | null;
  act_type_id: string;
  org_unit_id: string | null;
  content_html: string;
  content_json: Record<string, unknown> | null;
  content_mode: "rich_text" | "pdf" | "semantic" | "legacy_html" | "original_pdf";
  plain_text: string;
  status: MatterStatus;
  version: number;
  author_id: string;
  reviewed_by: string | null;
  published_at: string | null;
  is_erratum: boolean;
  act_number: string | null;
  act_year: number | null;
  act_date: string | null;
  responsible_name: string | null;
  responsible_role: string | null;
  responsible_id?: string | null;
  metadata?: Record<string, unknown> | null;
  review_reason?: string | null;
  publication_type: string;
  references_matter_id: string | null;
  created_at: string;
  updated_at: string;
  attachments: Attachment[];
}

export type DynamicFieldType =
  | "text"
  | "textarea"
  | "number"
  | "date"
  | "currency"
  | "cpf_cnpj"
  | "select"
  | "boolean";

export interface DynamicFieldDef {
  key: string;
  label: string;
  type: DynamicFieldType;
  required: boolean;
  placeholder?: string | null;
  help?: string | null;
  options?: string[];
}

export interface ActTypeConfig {
  number_required?: boolean;
  year_required?: boolean;
  date_required?: boolean;
  responsible_required?: boolean;
  allow_free_responsible?: boolean;
  title_pattern?: string | null;
  title_uppercase?: boolean;
  dynamic_fields?: DynamicFieldDef[];
}

export interface ActType {
  id: string;
  name: string;
  description: string | null;
  config?: ActTypeConfig | null;
}

export interface ActTypeAdmin extends ActType {
  is_active: boolean;
}

export interface Authority {
  id: string;
  name: string;
  role: string | null;
  org_unit_id: string | null;
  org_unit_name?: string | null;
  is_active: boolean;
  valid_from: string | null;
  valid_until: string | null;
  notes?: string | null;
  created_at?: string | null;
}

export interface OrgUnit {
  id: string;
  name: string;
  abbreviation: string | null;
  parent_id?: string | null;
  parent_name?: string | null;
}

export interface AuditEvent {
  id: string;
  action: string;
  description: string | null;
  created_at: string;
  extra_metadata: Record<string, unknown> | null;
}

export interface ApiError {
  detail: string;
}
